pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -316,60 +294,6 @@ protected synchronized boolean trySendMessagesToConsumers(ReadType readType, Lis
-                        lastSentPosition.getEntryId());
-                range = individuallySentPositions.firstRange();
-            }
-
-            if (range != null) {
-                // If the lowerBound is ahead of the last sent position,
-                // verify if there are any entries in-between.
-                if (range.lowerEndpoint().compareTo(lastSentPosition) <= 0 || managedLedger
-                        .getNumberOfEntries(com.google.common.collect.Range.openClosed(lastSentPosition,
-                                range.lowerEndpoint())) <= 0) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("[{}] Found a position range to last sent: {}", name, range);
-                    }
-                    Position newLastSentPosition = range.upperEndpoint();
-                    Position positionAfterNewLastSent = managedLedger
-                            .getNextValidPosition(newLastSentPosition);
-                    // sometime ranges are connected but belongs to different ledgers
-                    // so, they are placed sequentially
-                    // eg: (2:10..3:15] can be returned as (2:10..2:15],[3:0..3:15].
-                    // So, try to iterate over connected range and found the last non-connected range
-                    // which gives new last sent position.
-                    final Position lastConfirmedEntrySnapshot = managedLedger.getLastConfirmedEntry();
-                    if (lastConfirmedEntrySnapshot != null) {
-                        while (positionAfterNewLastSent.compareTo(lastConfirmedEntrySnapshot) <= 0) {
-                            if (individuallySentPositions.contains(positionAfterNewLastSent.getLedgerId(),
-                                    positionAfterNewLastSent.getEntryId())) {
-                                range = individuallySentPositions.rangeContaining(
-                                        positionAfterNewLastSent.getLedgerId(), positionAfterNewLastSent.getEntryId());
-                                newLastSentPosition = range.upperEndpoint();
-                                positionAfterNewLastSent = managedLedger.getNextValidPosition(newLastSentPosition);
-                                // check if next valid position is also deleted and part of the deleted-range
-                                continue;
-                            }
-                            break;
-                        }
-                    }
-
-                    if (lastSentPosition.compareTo(newLastSentPosition) < 0) {
-                        lastSentPosition = newLastSentPosition;
-                    }
-                    individuallySentPositions.removeAtMost(lastSentPosition.getLedgerId(),
-                            lastSentPosition.getEntryId());
-                }
-            }
-        }
 
         lastNumberOfEntriesProcessed = (int) totalEntriesProcessed;
```

Comment:
> ```java
> // line-281
> totalEntriesProcessed += entriesForConsumer.size();
> // line-282
> consumer.sendMessages(entriesForConsumer, batchSizes, batchIndexesAcks,
>        sendMessageInfo.getTotalMessages(),
>        sendMessageInfo.getTotalBytes(), sendMessageInfo.getTotalChunkedMessages(),
> ...
> // line-298
> this.lastNumberOfEntriesProcessed = (int) totalEntriesProcessed;
> ```

Since the method `consumer.sendMessages` will filter out entries due to the mechanism `consumer.pendingAcks.addPendingAckIfAllowed`, the variable `lastNumberOfEntriesProcessed` will be larger than expected

Reply:
> Since the method `consumer.sendMessages` will filter out entries due to the mechanism `consumer.pendingAcks.addPendingAckIfAllowed`, the variable `lastNumberOfEntriesProcessed` will be larger than expected

Thanks, very useful. I'll handle this in a follow-up PR.

Reply:
Tracked by #23420

Reply:
this is not a bug, explained in the comment on #23420